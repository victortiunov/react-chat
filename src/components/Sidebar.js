import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from 'material-ui/styles';
import Drawer from 'material-ui/Drawer';
import Divider from 'material-ui/Divider';
import TextField from 'material-ui/TextField';
import BottomNavigation, { BottomNavigationAction } from 'material-ui/BottomNavigation';
import MyChatsIcon from 'material-ui-icons/Person';
import GlobalIcon from 'material-ui-icons/Public';
import ChatList from './ChatList';
import NewChatButton from './NewChatButton';

const styles = theme => ({
  drawerPaper: {
    position: 'relative',
    height: '100%',
    width: 320,
  },
  drawerHeader: {
    ...theme.mixins.toolbar,
    paddingLeft: theme.spacing.unit * 3,
    paddingRight: theme.spacing.unit * 3,
  },
});

class Sidebar extends React.Component {
  static propTypes = {
    classes: PropTypes.objectOf(PropTypes.string).isRequired,
    activeChat: PropTypes.shape({
      _id: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      creator: PropTypes.shape({
        _id: PropTypes.string.isRequired,
      }).isRequired,
      members: PropTypes.arrayOf(PropTypes.object).isRequired,
    }),
    myChats: PropTypes.arrayOf(PropTypes.object).isRequired,
    allChats: PropTypes.arrayOf(PropTypes.object).isRequired,
    createChat: PropTypes.func.isRequired,
    setActiveChat: PropTypes.func.isRequired,
    isConnected: PropTypes.bool.isRequired,
  };

  static defaultProps = {
    activeChat: null,
  };

  constructor(props) {
    super(props);
    this.state = {
      activeTab: 0,
      searchTitle: '',
    };
  }

  getChats = () => {
    const searchTitle = this.state.searchTitle.toUpperCase();
    const chats = this.state.activeTab === 0 ? this.props.myChats : this.props.allChats;

    return chats.filter(chat => chat.title.toUpperCase().includes(searchTitle));
  };

  handleSearchTitleChange = (e) => {
    this.setState({ searchTitle: e.target.value });
  };

  handleTabChange = (e, value) => {
    this.setState({ activeTab: value, searchTitle: '' });
  };

  render() {
    const { activeTab, searchTitle } = this.state;
    const {
      classes, activeChat, createChat, setActiveChat, isConnected,
    } = this.props;

    return (
      <Drawer
        variant="permanent"
        classes={{
          paper: classes.drawerPaper,
        }}
      >
        <div className={classes.drawerHeader}>
          <TextField
            fullWidth
            margin="normal"
            value={searchTitle}
            placeholder="Search chats..."
            onChange={this.handleSearchTitleChange}
          />
        </div>
        <Divider />
        <ChatList
          activeChat={activeChat}
          chats={this.getChats()}
          setActiveChat={setActiveChat}
          disabled={!isConnected}
        />
        <NewChatButton createChat={createChat} disabled={!isConnected} />
        <BottomNavigation value={activeTab} onChange={this.handleTabChange} showLabels>
          <BottomNavigationAction label="My Chats" icon={<MyChatsIcon />} />
          <BottomNavigationAction label="Global" icon={<GlobalIcon />} />
        </BottomNavigation>
      </Drawer>
    );
  }
}

export default withStyles(styles)(Sidebar);
