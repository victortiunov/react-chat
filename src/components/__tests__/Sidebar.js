/* eslint-env jest */
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Sidebar from '../Sidebar';

jest.mock('../ChatList', () => () => 'ChatList');
jest.mock('../NewChatButton', () => () => 'NewChatButton');

const mockProps = {
  activeChat: {
    _id: '',
  },
  myChats: [],
  allChats: [],
  createChat: jest.fn(),
  setActiveChat: jest.fn(),
  isConnected: true,
};

describe('<Sidebar />', () => {
  it('render without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Sidebar {...mockProps} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders correctly', () => {
    const tree = renderer.create(<Sidebar {...mockProps} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
