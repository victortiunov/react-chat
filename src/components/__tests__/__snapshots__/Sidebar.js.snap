// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> renders correctly 1`] = `
<div
  className="MuiDrawer-docked-89"
>
  <div
    className="MuiPaper-root-100 MuiPaper-elevation0-102 MuiDrawer-paper-90 Sidebar-drawerPaper-87 MuiDrawer-paperAnchorLeft-91 MuiDrawer-paperAnchorDockedLeft-95"
  >
    <div
      className="Sidebar-drawerHeader-88"
    >
      <div
        aria-describedby={undefined}
        className="MuiFormControl-root-127 MuiFormControl-marginNormal-128 MuiFormControl-fullWidth-130"
        onBlur={[Function]}
        onFocus={[Function]}
      >
        <div
          className="MuiInput-root-131 MuiInput-fullWidth-138 MuiInput-formControl-132 MuiInput-underline-135"
        >
          <input
            aria-invalid={false}
            aria-required={false}
            autoComplete={undefined}
            autoFocus={undefined}
            className="MuiInput-input-139"
            defaultValue={undefined}
            disabled={false}
            id={undefined}
            name={undefined}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={undefined}
            onKeyUp={undefined}
            placeholder="Search chats..."
            readOnly={undefined}
            required={undefined}
            rows={undefined}
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <hr
      className="MuiDivider-root-145"
    />
    ChatList
    NewChatButton
    <div
      className="MuiBottomNavigation-root-149"
    >
      <button
        className="MuiButtonBase-root-157 MuiBottomNavigationAction-root-150 MuiBottomNavigationAction-selected-151"
        disabled={undefined}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiBottomNavigationAction-wrapper-153"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-160"
            color={undefined}
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
            />
          </svg>
          <span
            className="MuiBottomNavigationAction-label-154 MuiBottomNavigationAction-labelSelected-155"
          >
            My Chats
          </span>
        </span>
        <span
          className="MuiTouchRipple-root-166"
        />
      </button>
      <button
        className="MuiButtonBase-root-157 MuiBottomNavigationAction-root-150"
        disabled={undefined}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiBottomNavigationAction-wrapper-153"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-160"
            color={undefined}
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-1 17.93c-3.95-.49-7-3.85-7-7.93 0-.62.08-1.21.21-1.79L9 15v1c0 1.1.9 2 2 2v1.93zm6.9-2.54c-.26-.81-1-1.39-1.9-1.39h-1v-3c0-.55-.45-1-1-1H8v-2h2c.55 0 1-.45 1-1V7h2c1.1 0 2-.9 2-2v-.41c2.93 1.19 5 4.06 5 7.41 0 2.08-.8 3.97-2.1 5.39z"
            />
          </svg>
          <span
            className="MuiBottomNavigationAction-label-154"
          >
            Global
          </span>
        </span>
        <span
          className="MuiTouchRipple-root-166"
        />
      </button>
    </div>
  </div>
</div>
`;
